examples/drivencavity.py

```diff
@@ -1,60 +1,141 @@
          Element type (square/triangle/mixed).
        degree [2]
          Polynomial degree for velocity; the pressure space is one degree less.
        reynolds [1000]
          Reynolds number, taking the domain size as characteristic length.
+       strongbc [no]
+         Use strong boundary constraints
+       compatible [no]
+         Use compatible spaces and weakly imposed boundary conditions; requires
+         etype=square and strongbc=no
     '''
 
+    if compatible and (strongbc or etype != 'square'):
+        raise Exception(f'compatible mode requires square elements and weak boundary conditions')
+
     domain, geom = mesh.unitsquare(nelems, etype)
 
     ns = Namespace()
     ns.δ = function.eye(domain.ndims)
+    ns.ε = function.levicivita(2)
     ns.Σ = function.ones([domain.ndims])
     ns.Re = reynolds
+    ns.uwall = numpy.stack([domain.boundary.indicator('top'), 0])
     ns.x = geom
     ns.define_for('x', gradient='∇', normal='n', jacobians=('dV', 'dS'))
-    ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
-    ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1))
+    if not compatible:
+        ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
+        ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1)[1:])
+        ns.add_field('ψ', domain.basis('std', degree=2)[1:])
+    else:
+        ns.add_field(('u', 'v'), function.vectorize([
+            domain.basis('spline', degree=(degree, degree-1)),
+            domain.basis('spline', degree=(degree-1, degree))]))
+        ns.add_field(('p', 'q'), domain.basis('spline', degree=degree-1)[1:])
+        ns.add_field('ψ', domain.basis('spline', degree=degree)[1:])
     ns.σ_ij = '(∇_j(u_i) + ∇_i(u_j)) / Re - p δ_ij'
 
-    sqr = domain.boundary.integral('u_k u_k dS' @ ns, degree=degree*2)
-    sqr += domain.locate(ns.x, [(1,1)], weights=numpy.array([1.]), tol=1e-10).integral(ns.p**2)
-    cons = solver.optimize('u,p', sqr, droptol=1e-15)
+    res = domain.integral('∇_j(v_i) σ_ij dV' @ ns, degree=degree*2)
+    res += domain.integral('q ∇_k(u_k) dV' @ ns, degree=2*degree)
 
-    # project lid velocity separately to avoid Gibbs wiggles around boundary incompatibilities
-    sqr = domain.boundary['top'].integral('(u_0 - 1)^2 dS' @ ns, degree=degree*2)
-    lidcons = solver.optimize('u,', sqr, droptol=1e-15)
-    cons['u'] = numpy.choose(numpy.isnan(lidcons['u']), [lidcons['u'], cons['u']])
+    # strong enforcement of non-penetrating boundary conditions
+    sqr = domain.boundary.integral('(u_k n_k)^2 dS' @ ns, degree=degree*2)
+    cons = solver.optimize('u,', sqr, droptol=1e-15)
 
-    res = domain.integral('(∇_j(v_i) σ_ij + q ∇_k(u_k)) dV' @ ns, degree=degree*2)
-    with treelog.context('stokes'):
+    if strongbc:
+        # strong enforcement of tangential boundary conditions
+        sqr = domain.boundary['top,bottom'].integral('(u_0 - uwall_0)^2 dS' @ ns, degree=degree*2)
```

Comment:
I'm not sure if this an improvement, but you can combine the integrals using levi civita:

```python
sqr = domain.boundary.integral('(ε_ij n_i (u_j - uwall_j))^2 dS' @ ns, degree=degree*2)
```

Reply:
done